base/workerpool.jl

```diff
@@ -0,0 +1,85 @@
+# This file is a part of Julia. License is MIT: http://julialang.org/license
+
+
+type WorkerPool
+    channel::RemoteChannel{Channel{Int}}
+end
+
+
+isbusy(pool::WorkerPool) = !isready(pool.channel)
+
+
+"""
+    WorkerPool(workers)
+
+Create a WorkerPool from a vector of worker ids.
+"""
+function WorkerPool(workers::Vector{Int})
+
+    # Create a shared queue of workers...
+    pool = RemoteChannel(()->Channel{Int}(128)) 
+
+    # Check that workers are not already part of a pool...
+    check = () -> if :_worker_pool in names(Main)
+        error("Worker $(myid()) already in a WorkerPool!")
+    end
+    foreach(fetch, [@spawnat w check() for w in workers])
+
+    # Put each worker into the pool...
+    for w in workers
+        put!(pool, w)
+        @spawnat w global _worker_pool = pool
+    end
+
+    WorkerPool(pool)
+end
+
+
+"""
+    WorkerPool(n)
+
+Launches `n` workers using the in-built LocalManager.
+"""
+WorkerPool(n::Integer) = WorkerPool(addprocs(n))
+
+
+"""
+    WorkerPool()
+
+Equivalent to WorkerPool(CPU_CORES)
+"""
+WorkerPool() = WorkerPool(addprocs())
+
+
+"""
+    remotecall_fetch(f, pool::WorkerPool, args...)
+
+Call `f(args...)` on one of the workers in `pool`.
+"""
+function remotecall_fetch(f, pool::WorkerPool, args...)
+    l = (args...)->try f(args...) finally put!(_worker_pool, myid()) end
+    remotecall_fetch(l, take!(pool.channel), args...)
+end
+
+
+"""
+    default_worker_pool()
+
+Built-in WorkerPool (used by `remote(f)`).
+"""
+global _default_worker_pool = nothing
+function default_worker_pool()
```

Comment:
How about
 - doing away with a `_worker_pool`. Worker pools are created as and when required with a list of workers. `WorkerPool()` is equivalent to `WorkerPool(workers())` 
 - `_default_worker_pool` exists on pid 1. A worker is added to it here  https://github.com/JuliaLang/julia/blob/ca6f253c0c4c1c731621e02c1b00a6fe374fc473/base/multi.jl#L1063
 - `_default_worker_pool` on other workers are initialized via the `JoinPGRP` message  https://github.com/JuliaLang/julia/blob/ca6f253c0c4c1c731621e02c1b00a6fe374fc473/base/multi.jl#L46
 - On other workers the channel referred by it is the same one as on pid 1